File: packages/health_connector_hc_android/android/src/main/kotlin/com/phamtunglam/health_connector_hc_android/mappers/health_record_mappers/ExerciseSessionEventMapper.kt
package com.phamtunglam.health_connector_hc_android.mappers.health_record_mappers

import androidx.health.connect.client.records.ExerciseLap
import androidx.health.connect.client.records.ExerciseSegment
import androidx.health.connect.client.units.Length
import com.phamtunglam.health_connector_hc_android.pigeon.ExerciseSessionLapEventDto
import com.phamtunglam.health_connector_hc_android.pigeon.ExerciseSessionSegmentEventDto
import java.time.Instant

/**
 * Converts a Health Connect [ExerciseSegment] to [ExerciseSessionSegmentEventDto].
 */
internal fun ExerciseSegment.toDto(): ExerciseSessionSegmentEventDto =
    ExerciseSessionSegmentEventDto(
        startTime = startTime.toEpochMilli(),
        endTime = endTime.toEpochMilli(),
        segmentType = segmentType.toExerciseSegmentTypeDto(),
        repetitions = if (repetitions > 0) repetitions.toLong() else null,
    )

/**
 * Converts a Health Connect [ExerciseLap] to [ExerciseSessionLapEventDto].
 */
internal fun ExerciseLap.toDto(): ExerciseSessionLapEventDto = ExerciseSessionLapEventDto(
    startTime = startTime.toEpochMilli(),
    endTime = endTime.toEpochMilli(),
    distanceMeters = length?.inMeters,
)

/**
 * Converts [ExerciseSessionSegmentEventDto] to Health Connect [ExerciseSegment].
 */
internal fun ExerciseSessionSegmentEventDto.toHealthConnect(): ExerciseSegment = ExerciseSegment(
    startTime = Instant.ofEpochMilli(startTime),
    endTime = Instant.ofEpochMilli(endTime),
    segmentType = segmentType.toHealthConnect(),
    repetitions = repetitions?.toInt() ?: 0,
)

/**
 * Converts [ExerciseSessionLapEventDto] to Health Connect [ExerciseLap].
 */
internal fun ExerciseSessionLapEventDto.toHealthConnect(): ExerciseLap = ExerciseLap(
    startTime = Instant.ofEpochMilli(startTime),
    endTime = Instant.ofEpochMilli(endTime),
    length = distanceMeters?.let { Length.meters(it) },
)

File: packages/health_connector_hc_android/lib/src/mappers/health_record_mappers/exercise/exercise_session_event_mapper.dart
import 'package:health_connector_core/health_connector_core_internal.dart'
    show
        ExerciseSessionLapEvent,
        ExerciseSessionSegmentEvent,
        ExerciseSessionEvent,
        Length,
        ExerciseSessionStateTransitionEvent,
        ExerciseSessionMarkerEvent;
import 'package:health_connector_hc_android/src/mappers/health_record_mappers/exercise/exercise_segment_type_mapper.dart';
import 'package:health_connector_hc_android/src/pigeon/health_connector_hc_android_api.g.dart'
    show
        ExerciseSessionLapEventDto,
        ExerciseSessionSegmentEventDto,
        ExerciseSessionEventDto;
import 'package:meta/meta.dart' show internal;

/// Extension to convert [ExerciseSessionEvent] to [ExerciseSessionEventDto].
@internal
extension ExerciseSessionEventToDto on ExerciseSessionEvent {
  ExerciseSessionEventDto toDto() {
    return switch (this) {
      ExerciseSessionLapEvent(
        :final startTime,
        :final endTime,
        :final distance,
      ) =>
        ExerciseSessionLapEventDto(
          startTime: startTime.millisecondsSinceEpoch,
          endTime: endTime.millisecondsSinceEpoch,
          distanceMeters: distance?.inMeters,
        ),
      ExerciseSessionSegmentEvent(
        :final startTime,
        :final endTime,
        :final segmentType,
        :final repetitions,
      ) =>
        ExerciseSessionSegmentEventDto(
          startTime: startTime.millisecondsSinceEpoch,
          endTime: endTime.millisecondsSinceEpoch,
          segmentType: segmentType.toDto(),
          repetitions: repetitions,
        ),
      ExerciseSessionStateTransitionEvent _ => throw UnsupportedError(
        '$ExerciseSessionStateTransitionEvent is not supported on '
        'Android Health Connect. Use $ExerciseSessionLapEvent and '
        '$ExerciseSessionSegmentEvent instead.',
      ),
      ExerciseSessionMarkerEvent _ => throw UnsupportedError(
        '$ExerciseSessionMarkerEvent is not supported on '
        'Android Health Connect. Use $ExerciseSessionLapEvent and '
        '$ExerciseSessionSegmentEvent instead.',
      ),
    };
  }
}

/// Extension to convert [ExerciseSessionEventDto] to [ExerciseSessionEvent].
@internal
extension ExerciseSessionEventDtoToDomain on ExerciseSessionEventDto {
  ExerciseSessionEvent toDomain() {
    return switch (this) {
      ExerciseSessionLapEventDto(
        :final startTime,
        :final endTime,
        :final distanceMeters,
      ) =>
        ExerciseSessionLapEvent(
          startTime: DateTime.fromMillisecondsSinceEpoch(
            startTime,
            isUtc: true,
          ),
          endTime: DateTime.fromMillisecondsSinceEpoch(endTime, isUtc: true),
          distance: distanceMeters != null
              ? Length.meters(distanceMeters)
              : null,
        ),
      ExerciseSessionSegmentEventDto(
        :final startTime,
        :final endTime,
        :final segmentType,
        :final repetitions,
      ) =>
        ExerciseSessionSegmentEvent(
          startTime: DateTime.fromMillisecondsSinceEpoch(
            startTime,
            isUtc: true,
          ),
          endTime: DateTime.fromMillisecondsSinceEpoch(endTime, isUtc: true),
          segmentType: segmentType.toDomain(),
          repetitions: repetitions,
        ),
    };
  }
}
